make validate debug-only in Device ctr

bdhirsh opened the pull request:
This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47854 make validate debug-only in Device ctr**

Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

bdhirsh committed:
[TEST] make validate debug-only in Device copy ctr

[ghstack-poisoned]

bdhirsh committed:
Update on "[TEST] make validate debug-only in Device copy ctr"




[ghstack-poisoned]

bdhirsh committed:
Update on "[TEST] make validate debug-only in Device copy ctr"




[ghstack-poisoned]

bdhirsh committed:
Update on "[TEST] make validate debug-only in Device copy ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[TEST] make validate debug-only in Device copy ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[Test] make validate debug-only in Device ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[Test] make validate debug-only in Device ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[Test] make validate debug-only in Device ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[Test] make validate debug-only in Device ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "[Test] make validate debug-only in Device ctr"


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh commented on torch/csrc/Device.cpp:
```diff
@@ -65,9 +65,6 @@ PyObject *THPDevice_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
-      TORCH_CHECK(device_index >= 0, "Device index must not be negative");
     }
     at::Device device(as_device.type(), device_index);
     return THPDevice_New(device);
```
I put a little effort into fixing the (existing) clang-tidy errors, but we might just have to ignore/suppress them. clang-tidy complains about global non-const objects, but it looks like PyBind expects these objects to be non-const (both the `PyTypeObject` constructor arguments and the call to `PyType_Ready()`). We could add the const-ness and cast it away, but "hiding" const-ness seems like a bad way to appease clang-tidy.

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

ezyang commented on torch/csrc/Device.cpp:
```diff
@@ -65,9 +65,6 @@ PyObject *THPDevice_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
       device_index = r.toInt64(1);
-      // -1 is allowed in ATen/C++, to mean the default device, but not in
-      // Python.
-      TORCH_CHECK(device_index >= 0, "Device index must not be negative");
```
What's going on here? This test doesn't seem harmful. Is this for lining up the C++ and Python behavior? A little bit of divergence is probably OK; it's better to allow less than to allow more (you can make things more relaxed in the future, but not vice versa)

bdhirsh replied:
Yeah. I was a little torn because of exactly what you just said- taking it out now means we can't put it back later.

But I figured that the target state we want is to have consistent behavior between the python and C++. It seems a little weird to randomly throw an error in python and have the exact same logic pass in C++.

You think it's better to keep in for now?

ezyang reviewed (approved):
Short and sweet.

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]

bdhirsh committed:
Update on "make validate debug-only in Device ctr"

This PR makes the validation check in the `Device` constructor a debug-only assert.

**Background**
Tensors know about the `Device` that they run on, e.g. `cpu` or `cuda:0`. Devices have a DeviceIndex that's only really applicable for cuda/hip, and is effectively ignored for other devices. Previously, we ran "some" validation upon device construction- basically maintain the invariant that `DeviceIndex == -1` for `cpu` devices and `>= 0` for non-cpu devices.

This assertion is technically redundant for cuda/hip devices, and unnecessary for cpu devices.

For cuda/hip, we perform a later check when we try to switch to the specific DeviceIndex anyway. Removing this first check means that we fail slightly later, e.g, if someone creates a `cuda:-1` device we will fail at usage time rather than construction time (this check doesn't cover everything anyway, e.g. indexes that are out of range).

For cpu devices, there isn't really a purpose in maintaining the invariant that `DeviceIndex == -1`, as we never use it anywhere. TBC, that means that code like this runs fine (it used to fail): `torch.device("cpu", 5)`.

**Purpose**
As Ed pointed out in a [previous PR](https://github.com/pytorch/pytorch/pull/44322), this check has a noticeable impact on the wall-time of several microbenchmarks.

Just as a note, the reason is likely because we repeatedly construct Device objects several times in hot-path code. I made an effort to remove the number of Device objects that we construct [here](https://github.com/pytorch/pytorch/pull/47894), but had some trouble finding a noticeable perf improvement.

Framework overhead benchmarks show ~1-10% wall-time improvements, depending on the microbenchmark: https://www.internalfb.com/intern/aibench/details/1862369817/


Differential Revision: [D25003113](https://our.internmc.facebook.com/intern/diff/D25003113)

[ghstack-poisoned]
